sql/src/main/java/org/apache/druid/sql/calcite/aggregation/builtin/GreatestSqlAggregator.java

```diff
@@ -0,0 +1,122 @@
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.sql.calcite.aggregation.builtin;
+
+import org.apache.calcite.rel.core.AggregateCall;
+import org.apache.calcite.sql.SqlAggFunction;
+import org.apache.calcite.sql.SqlFunctionCategory;
+import org.apache.calcite.sql.SqlKind;
+import org.apache.calcite.sql.type.OperandTypes;
+import org.apache.calcite.sql.type.ReturnTypes;
+import org.apache.druid.java.util.common.ISE;
+import org.apache.druid.math.expr.ExprMacroTable;
+import org.apache.druid.query.aggregation.AggregatorFactory;
+import org.apache.druid.query.aggregation.DoubleMaxAggregatorFactory;
+import org.apache.druid.query.aggregation.LongMaxAggregatorFactory;
+import org.apache.druid.query.aggregation.PostAggregator;
+import org.apache.druid.query.aggregation.post.DoubleGreatestPostAggregator;
+import org.apache.druid.query.aggregation.post.FieldAccessPostAggregator;
+import org.apache.druid.query.aggregation.post.LongGreatestPostAggregator;
+import org.apache.druid.segment.column.ValueType;
+import org.apache.druid.sql.calcite.aggregation.Aggregation;
+import org.apache.druid.sql.calcite.planner.Calcites;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Calcite integration class for Greatest post aggregators of Long & Double types.
+ * It applies Max aggregators over the provided fields/expressions & combines their results via Field access post aggregators.
+ */
+public class GreatestSqlAggregator extends MultiColumnSqlAggregator
+{
+  private static final SqlAggFunction FUNCTION_INSTANCE = new GreatestSqlAggFunction();
+  private static final String NAME = "GREATEST";
+
+  @Override
+  public SqlAggFunction calciteFunction()
+  {
+    return FUNCTION_INSTANCE;
+  }
+
+  @Override
+  Aggregation getAggregation(
+      String name,
+      AggregateCall aggregateCall,
+      ExprMacroTable macroTable,
+      List<FieldInfo> fieldInfoList
+  )
+  {
+    final ValueType valueType = Calcites.getValueTypeForSqlTypeName(aggregateCall.getType().getSqlTypeName());
+    List<AggregatorFactory> aggregators = new ArrayList<>();
+    List<PostAggregator> postAggregators = new ArrayList<>();
+
+    // Create Max aggregator factories for provided fields & corresponding field access post aggregators.
+    int id = 0;
+    for (FieldInfo fieldInfo : fieldInfoList) {
+      String prefixedName = Calcites.makePrefixedName(name, String.valueOf(id++));
+      postAggregators.add(new FieldAccessPostAggregator(null, prefixedName));
+      switch (valueType) {
+        case LONG:
+          aggregators.add(new LongMaxAggregatorFactory(prefixedName, fieldInfo.fieldName, fieldInfo.expression, macroTable));
+          break;
+        case FLOAT:
```

Comment:
The min and max aggregators use a different aggregator for float vs double. Any reason you chose to have both use the same aggregator factory

Reply:
The only reason I used "DoublePostAggregators" for float aggregators is that there are no "FloatPostAggregators" in druid. Even though float min/max aggregators are there, post aggregators specific to that type are missing.
I believe this is the only way to use decimal post aggregators in druid JSON query as of now.
Please review & update if there's anything I'm missing here.